app/controllers/api/v1x3/progress_messages_controller.rb

```diff
@@ -0,0 +1,18 @@
+module Api
+  module V1x3
+    class ProgressMessagesController < ApplicationController
+      include Mixins::IndexMixin
+
+      SOURCE_TYPES = {"Order" => "order_id", "OrderItem" => "order_item_id"}.freeze
+
+      def index
+        source_type = params[:source_type] || "OrderItem"
+
+        # The source_type should also match with request path
+        raise Catalog::InvalidParameter unless SOURCE_TYPES.include?(source_type) && request.path.include?(source_type.constantize.table_name)
```

Comment:
The logic is hard to understand. We can easily determine source_type by checking whether exists `params[:order_id]` or `params[:order_item_id]`, otherwise raise an error.